libevmasm/SemanticInformation.cpp

```diff
@@ -278,9 +310,45 @@ bool SemanticInformation::invalidatesStorage(Instruction _instruction)
 	case Instruction::CREATE:
 	case Instruction::CREATE2:
 	case Instruction::SSTORE:
-		return true;
+		return yul::SideEffects::Write;
+
+	case Instruction::SLOAD:
+	case Instruction::STATICCALL:
+		return yul::SideEffects::Read;
+
 	default:
-		return false;
+		return yul::SideEffects::None;
+	}
+}
+
+yul::SideEffects::Effect SemanticInformation::otherState(Instruction _instruction)
+{
+	switch (_instruction)
+	{
+	case Instruction::CALL:
+	case Instruction::CALLCODE:
+	case Instruction::DELEGATECALL:
+	case Instruction::CREATE:
+	case Instruction::CREATE2:
+	case Instruction::SELFDESTRUCT:
+		// Strictly speaking, log0, .., log4 writes to the state, but the EVM cannot read it, so they
+		// are just marked as having 'other side effects.'
+		return yul::SideEffects::Write;
+
+	case Instruction::EXTCODESIZE:
+	case Instruction::EXTCODEHASH:
+	case Instruction::RETURNDATASIZE:
+	case Instruction::BALANCE:
+	case Instruction::SELFBALANCE:
+	case Instruction::RETURNDATACOPY:
+	case Instruction::EXTCODECOPY:
+	case Instruction::STATICCALL:
```

Comment:
If we say that `returndatasize` reads, then `staticcall` should write.

Reply:
Can you add a test (in the other pr) thaht would try to pull out a `returndatasize` from a loop that contains a `staticcall`?

Reply:
I changed `staticall` to write and added a test: https://github.com/ethereum/solidity/pull/9312/files#diff-f1daf8a43b1098ed8941a35bf1c4f973

I think `staticall` shouldn't affect the other opcodes that reads state such as `extcodehash, balance` etc. Currently, these opcodes won't be moved. Do you think we should fine tune the optimizer for `staticall`?

Reply:
we could split `otherstate` even more if you ilke. I'm a bit worried about performance, though.